lib/relation-definition.js

```diff
@@ -532,37 +532,97 @@ function lookupModel(models, modelName) {
   }
 }
 
-function lookupModelTo(modelFrom, modelTo, params, singularize) {
-  if ('string' === typeof modelTo) {
+/*
+ * @param {Object} modelFrom Instance of the 'from' model
+ * @param {Object|String} modelToRef Reference to Model object to which you are
+ *  creating the relation: model instance, model name, or name of relation to model.
+ * @param {Object} params The relation params
+ * @param {Boolean} singularize Whether the modelToRef should be singularized when
+ *  looking-up modelTo
+ * @return {Object} modelTo Instance of the 'to' model
+ */
+function lookupModelTo(modelFrom, modelToRef, params, singularize) {
+  var modelTo;
+
+  if (typeof modelToRef !== 'string') {
+    // modelToRef might already be an instance of model
+    modelTo = modelToRef;
+  } else {
+    // lookup modelTo based on relation params and modelToRef
     var modelToName;
-    params.as = params.as || modelTo;
-    modelTo = params.model || modelTo;
-    if (typeof modelTo === 'string') {
-      modelToName = (singularize ? i8n.singularize(modelTo) : modelTo).toLowerCase();
-      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName) || modelTo;
-    }
+    modelTo = params.model || modelToRef; // modelToRef might be modelTo name
+
     if (typeof modelTo === 'string') {
-      modelToName = (singularize ? i8n.singularize(params.as) : params.as).toLowerCase();
-      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName) || modelTo;
+      // lookup modelTo by name
+      modelToName = modelTo;
+      modelToName = (singularize ? i8n.singularize(modelToName) : modelToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
-    if (typeof modelTo !== 'function') {
-      throw new Error(g.f('Could not find "%s" relation for %s', params.as, modelFrom.modelName));
+
+    if (!modelTo) {
+      // lookup by modelTo name was not successful. Now looking-up by relationTo name
+      var relationToName = params.as || modelToRef;  // modelToRef might be relationTo name
+      modelToName = (singularize ? i8n.singularize(relationToName) : relationToName).toLowerCase();
+      modelTo = lookupModel(modelFrom.dataSource.modelBuilder.models, modelToName);
     }
   }
+  if (typeof modelTo !== 'function') {
+    throw new Error(g.f('Could not find relation %s for model %s', params.as, modelFrom.modelName));
+  }
   return modelTo;
 }
 
-/*!
+/*
+ * @param {Object} params The relation params
+ * @param {String} relationName The relation name
+ * NOTE: settleRelationAs() mutates the params object
+ */
+function settleRelationAs(params, relationName) {
+  if (typeof relationName === 'string') {
+    params.as = params.as || relationName;
+  }
+}
```

Comment:
this is the new helper to assign relation's `params.as` outside of `lookupModelTo`. It is used everywhere `lookupModelTo` is used.